contents/data/nosql/NoSQL.md

```diff
@@ -0,0 +1,436 @@
+  popularity: 9.7,
+  ...
+}
+```
+
+<br>
+<box type="info" light>
+Note that while many NoSQL databases provide a JSON interface to interact with the data,
+the underlying storage implementation may be different for performance reasons.<br>
+For example, MongoDB stores documents in BSON (json encoded in binary). 
+</box>
+
+<br>
+
+### Basic CRUD operation examples
+
+Let's skip the trivial details of database setup which would differ across different database technologies,
+and look at how you would interact with NoSQL databases.
+
+Unlike SQL, where database interaction is achieved through the different dialects of the query language,
+most interaction with NoSQL databases is achieved in a simple and intuitive _object oriented manner_,
+and JSON-like queries.
+
+Let's get back to the above example. Say you already have the _customers_ collection set up and ready to go.
+A potential new user sends a POST request to register a new account, and your server would then simply call
+the following mongoDB code to store the details after some processing.
+
+
+```js
+// Example insertion of a customer document into the _customers_ collection
+db.customers.insertOne({
+  username: "panda",
+  password: hashedPassword,
+  email: "panda@pandas.com"
+})
+```
+
+That was a rather simple use case, but you get the idea. Let's get into something more powerful.
+
+<div id="schemaless">
+
+Suppose our _items_ collection is structured like so:
+</div>
+
+```js
+[
+  {
+    type: "book",
+    title: "about pandas",
+    price: 20,
+    popularity: 9.7,
+    author: "panda1",
+    ...
+  },
+  {
+    title: "more pandas",
+    author: "panda2",
+  },
+  ...
+]
+```
+
+And a customer filters through your catalogue with a price of less than 30, and a popularity with more than 8.
+Furthermore, he / she sorts the items lexicographically by the price, then by descending order of the title.
+
+To retrieve that data, your backend would make a query like so:
+```js
+db.items.find({
+  price: { $lt: 30 },
+  popularity: { $gt: 8 }
+}).sort({
+  // Here 1 means ascending order, and -1 means descending
+  price: 1,
+  title: -1
+})
+```
+
+Here, the code simply provides a query JSON object matching the structure of an item document.
+
+Specifically, items with a price of less than 30 are filtered with the `$lt: 30` key-value pair,
+and items with a popularity of more than 8 are filtered with the `$gt: 8`.<br>
+Subsequently, a
+<popover header="Cursor :fas-gamepad:">
+<template slot="content">
+Think of it as a controller with a button which you can repeatedly press to retrieve the next document,
+with several smaller buttons to change the next document retrieved!
+</template>"cursor"</popover>
+pointing to the retrieved subset of documents is returned, which undergoes a `sort`
+operation by the respective fields of `price` and `title`.
+
+CRUD operations with NoSQL databases are just as powerful as traditional SQL databases, and in some
+cases even more succinct.
+
+<br>
```

Comment:
What's the point you are trying to make in this section? If the point is the following, perhaps the description above need not go into so much low level details?
>CRUD operations with NoSQL databases are just as powerful as traditional SQL databases, and in some
cases even more succinct.